src/ui/public/vis/request_handlers/courier.js

```diff
@@ -35,15 +37,22 @@ const CourierRequestHandlerProvider = function (Private, courier, timefilter) {
             };
 
             searchSource.rawResponse = resp;
-            resolve(resp);
+
+            resolve(_.cloneDeep(resp));
           }).catch(e => reject(e));
 
           courier.fetch();
         } else {
-          resolve(searchSource.rawResponse);
+          resolve(_.cloneDeep(searchSource.rawResponse));
         }
-
-
+      }).then(async resp => {
+        for (let i = 0; i < vis.aggs.length; i++) {
+          const agg = vis.aggs[i];
+          if (!agg.type || !agg.type.postFlightRequest) continue;
+          const nestedSearchSource = new SearchSource().inherits(searchSource);
+          resp = await agg.type.postFlightRequest(vis.aggs, agg, searchSource.get('aggs')(), resp, nestedSearchSource);
```

Comment:
I'm not a 100% convinced this `postFlightRequest` is the correct abstraction.

This only works for terms, where we can conceive of something like "other", where this configuration requires more calls. But this wouldn't transfer to other aggregations, if a postflight makes sense. This also bleeds into courier in kind of a weird way. Also, the toDSL functionality is now broken for Terms. The configuration of terms and all its outputs (dsl, responses, ...) no longer matches the actual result.

Could we flip this? Instead of introducing a `postFlightRequest`, do preliminary requests at the _beginning_.

- We could do something similar how the `geo_hash` aggregation rewrites the query in `getRequestAggs` (https://github.com/thomasneirynck/kibana/blob/05cbb929658590024e231a0791311c8b859e194f/src/ui/public/agg_types/buckets/geo_hash.js#L106-L106). This would mean of course that we make this method (getRequestAggs) async.
- If the above doesn't work, we could introduce an async `preFlight` method that rewrites the query-object. This could do all the intermediate calls required to produce the final query.

In either case, I think it would preserve the contract of the "Terms" configuration better.

Reply:
i don't think that is possible, we discussed this with @nreese and found few reasons why to do the postFlight instead of preFlight ... i think the main one was that i need the response for this to work
- i can not construct correct filters i need for the 'other' bucket construction if i don't have the response from original query back as this could be nested 
- we are updating the actual response with fake buckets, which can't really happen in pre-flight

Reply:
The idea would be that you make the original request in the preflight, and collect the other responses in a similar fashion as here, except for the last one. Your last DSL will then be the correct one, matching the terms-config. Right now, we do the first request (not matching the terms config) and stringing together all the subsequent ones in the postflight.